rwt_teleop/euslisp/pr2-realtime-ik.l

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env roseus
+(require :pr2-interface "package://pr2eus/pr2-interface.l")
+
+(ros::roseus "pr2_realtime_ik")
+
+(pr2-init)
+(setq *robot* *pr2*)
+(setq larm-tgt (make-coords :pos #f(500 300 1000)))
+(setq rarm-tgt (make-coords :pos #f(500 -300 1000)))
+(setq head-tgt (make-coords))
+
+(objects (list *robot* larm-tgt rarm-tgt))
+
+(defclass realtime-ik-class
+  :super propertied-object
+  :slots (realtime-ik-))
+(defmethod realtime-ik-class
+  (:init ()
+	 (ros::subscribe "master_larm_pose" geometry_msgs::PoseStamped #'send self :larm-cb)
+	 (ros::subscribe "master_rarm_pose" geometry_msgs::PoseStamped #'send self :rarm-cb)
+	 (ros::subscribe "master_head_pose" geometry_msgs::PoseStamped #'send self :head-cb)
+   (setq realtime-ik- t)
+   (ros::advertise-service "~/start" std_srvs::Empty #'send self :start-realtime-ik)
+   (ros::advertise-service "~/stop" std_srvs::Empty #'send self :stop-realtime-ik)
+	 )
+  (:larm-cb (msg)
+	    (setq raw-coords (ros::tf-pose->coords (send msg :pose)))
+	    (setq llim #f( 400  -200  500))
+	    (setq ulim #f(1500  1500 1500))
+	    (setq safe-pos (vmin (vmax (send raw-coords :pos) llim) ulim))
+	    (setq larm-tgt (make-coords :pos safe-pos :rot (send raw-coords :rot))))
+  (:rarm-cb (msg)
+	    (setq raw-coords (ros::tf-pose->coords (send msg :pose)))
+	    (setq llim #f( 400 -1500  500))
+	    (setq ulim #f(1500   200 1500))
+	    (setq safe-pos (vmin (vmax (send raw-coords :pos) llim) ulim))
+	    (setq rarm-tgt (make-coords :pos safe-pos :rot (send raw-coords :rot))))
+  (:head-cb (msg)
+	    (setq head-tgt (ros::tf-pose->coords (send msg :pose))))
+  (:start-realtime-ik (req)
+    (ros::ros-info "start-realtime-ik")
+    (setq realtime-ik- t)
+    (send req :response))
+  (:stop-realtime-ik (req)
+    (ros::ros-info "stop-realtime-ik")
+    (setq realtime-ik- nil)
+    (send req :response))
+  (:realtime-ik-p ()
+    realtime-ik-)
+)
+
+
+(setq realtime-ik-instance (instance realtime-ik-class :init))
+
+(defun run ()
+  (send *robot* :reset-pose)
+  (send *robot* :head :neck-p :joint-angle 60)
+  (send *robot* :torso_lift_joint :joint-angle 200)
+  (send *ri* :angle-vector (send *robot* :angle-vector) 3000)
+  (send *ri* :wait-interpolation)
+
+  (ros::rate 5)
+  (while (ros::ok)
+    (when (send realtime-ik-instance :realtime-ik-p)
```

Comment:
use if to avoid macro expansion and gc running

Reply:
```lisp
(defun foo-3 (n)
  (let ((i 0))
    (while (< i n) (when t'(1 2 3 4 5) (setq i (1+ i))))
    nil))

(defun foo-4 (n)
  (let ((i 0))
    (while (< i n) (if t (progn '(1 2 3 4 5) (setq i (1+ i)))))
    nil))
```

```
1.irteusgl$ (bench (foo-3 10e6))
;; time -> 6.9292[s]
nil
2.irteusgl$ (bench (foo-4 10e6))
;; time -> 1.17928[s]
nil
```

Reply:
Thank you for your advice.

I didn't know how macro expansion and gc running take time.

That's good to hear.